deep-learning/src/main/scala/com/microsoft/ml/spark/onnx/ONNXModel.scala

```diff
@@ -0,0 +1,496 @@
+    }
+  }
+
+  /**
+   * Returns true if the two data types are compatible. They are compatible if they share the same "shape", and
+   * 1. The element types from both sides are numeric types, or
+   * 2. The element types from both sides are the same.
+   */
+  @tailrec
+  private def compatible(from: DataType, to: DataType): Boolean = {
+    (from, to) match {
+      case (VectorType, right: ArrayType) =>
+        compatible(DoubleType, right.elementType)
+      case (left: ArrayType, right: ArrayType) =>
+        compatible(left.elementType, right.elementType)
+      case (_: NumericType, _: NumericType) => true
+      case (fromDataType, toDataType) => fromDataType == toDataType
+    }
+  }
+}
+
+class ONNXModel(override val uid: String)
+  extends Transformer
+    with ComplexParamsWritable
+    with ONNXModelParams
+    with Wrappable
+    with BasicLogging {
+
+  import ONNXModel._
+
+  override protected lazy val pyInternalWrapper = true
+
+  logClass()
+
+  def this() = this(Identifiable.randomUID("ONNXModel"))
+
+  @transient
+  lazy val modelInput: Map[String, NodeInfo] = {
+    using(initializeOrt(getModelPayload)) {
+      session => session.getInputInfo.asScala.toMap
+    }.get
+  }
+
+  @transient
+  lazy val modelOutput: Map[String, NodeInfo] = {
+    using(initializeOrt(getModelPayload)) {
+      session => session.getOutputInfo.asScala.toMap
+    }.get
+  }
+
+  override def transform(dataset: Dataset[_]): DataFrame = logTransform {
+    val inputSchema = dataset.schema
+    val outputSchema = transformValidateSchema(inputSchema, includePostProcessFields = false)
+
+    implicit val enc: Encoder[Row] = RowEncoder(
+      StructType(outputSchema.map(f => StructField(f.name, ArrayType(f.dataType))))
+    )
+
+    val modelPayloadBC = dataset.sparkSession.sparkContext.broadcast(this.getModelPayload)
+
+    // The cache call is a workaround for GH issue 1075:
+    //  https://github.com/Azure/mmlspark/issues/1075
+    val batchedDF = getMiniBatcher.transform(dataset).cache()
+    val (coerced, feedDict) = coerceBatchedDf(batchedDF)
+
+    val outputDf = coerced.mapPartitions {
+      rows: Iterator[Row] =>
+        val session = initializeOrt(modelPayloadBC.value)
+        applyModel(session, feedDict, getFetchDict, inputSchema)(rows)
+    }
+
+    // The cache call is a workaround for GH issue 1075:
+    //  https://github.com/Azure/mmlspark/issues/1075
+    val cacheAttempted = if (outputDf.isStreaming) outputDf else outputDf.cache()
+
+    val flattenedDF = new FlattenBatch().transform(cacheAttempted)
+
+    (softMaxTransform _ andThen argMaxTransform) (flattenedDF)
+  }
+
+  private def softMaxTransform(input: DataFrame): DataFrame = {
+    this.getSoftMaxDict.foldLeft(input) {
+      case (df, (input, output)) =>
+        val softmaxCol = df.schema(input).dataType match {
+          case ArrayType(_: NumericType, _) =>
+            val softmaxUdf = UDFUtils.oldUdf({
+              array: Seq[Double] =>
```

Comment:
Will this ever be Seq[Float]?

Reply:
Yes. 5 lines below where the UDF is called, we cast the input to `Array[Double]`, then call the UDF, thus ensuring the input is `Seq[Double]` type.
```
softmaxUdf(col(input).cast(ArrayType(DoubleType)))
```